flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -351,6 +353,7 @@ public Task(
 		this.taskManagerConfig = Preconditions.checkNotNull(taskManagerConfig);
 
 		this.metrics = metricGroup;
+		metricGroup.gauge(MetricNames.IS_BACKPRESSURED, new BackPressureGauge(this));
```

Comment:
why not just `metricGroup.gauge(MetricNames.IS_BACKPRESSURED, this::isBackPressured);`?

Also, since this effectively leaks a reference to an object under construction, have you verified that it is safe to call `isBackpressured` without the constructed having finished?

Reply:
What do you think about exposing it as a numeric value instead? (0/1)

Reply:
I'm not sure what "0.3" of back pressure should mean? Task is back pressured, or it's not. Unless 0.3 would mean that task is back pressured 30% of the time? But such kind of aggregation maybe should be performed on some upper level (if at all).

Reply:
I literally meant 0 or 1; 0 for not being back-pressured, 1 for being back-pressured.

But this would also be a mapping reporters could do; we don't have to deal with these details here.

Reply:
Oh, OK, I get it now. I don't mind having an integer vs boolean. 

What's the reason that you prefer an integer? That it's easier to aggregate in systems like Grafana?

Reply:
purely because some metric backends only support numbers, but again we shouldn't worry about this here; that's a reporter detail.

Reply:
> If you're going for a Gauge, why not just `metricGroup.gauge(MetricNames.IS_BACKPRESSURED, this::isBackPressured);`?
> 
> Also, since this effectively leaks a reference to an object under construction, have you verified that it is safe to call `isBackpressured` without the constructor having finished?
Under what circumstances would call `isBackpressured` without the constructor having finished？

Reply:
> org.apache.flink.runtime.taskexecutor.TaskExecutor#submitTask will be better.

I think that would be a safer, however a bit more ugly solution. I would go with that option for now, especially that it looks like it's currently broken (`consumableNotifyingPartitionWriters` is being initialised after  this line).

and +1 for  @zentol's suggestion of removing `BackPressureGauge` in favour of a method reference.

Reply:
Ok, I'll remove BackPressureGauge and move register action to TaskExecutor#submitTask.

Reply:
> Under what circumstances would call isBackpressured without the constructor having finished？

@jinglining The moment `gauge()` returns the metric is accessible for the metric system, and may call it before the rest of the constructor has run.

Reply:
> I think that would be a safer, however a bit more ugly solution.

It may be fine to register the metric at the very end of the constructor.